dspace-server-webapp/src/main/java/org/dspace/app/rest/security/OrcidLoginFilter.java

```diff
@@ -0,0 +1,122 @@
+        throws AuthenticationException {
+
+        if (!OrcidAuthentication.isEnabled()) {
+            throw new ProviderNotFoundException("Orcid login is disabled.");
+        }
+
+        return authenticationManager.authenticate(new DSpaceAuthentication());
+
+    }
+
+    @Override
+    protected void successfulAuthentication(HttpServletRequest req, HttpServletResponse res, FilterChain chain,
+        Authentication auth) throws IOException, ServletException {
+
+
+        DSpaceAuthentication dSpaceAuthentication = (DSpaceAuthentication) auth;
+
+        log.debug("Orcid authentication successful for EPerson {}. Sending back temporary auth cookie",
+            dSpaceAuthentication.getName());
+
+        restAuthenticationService.addAuthenticationDataForUser(req, res, dSpaceAuthentication, true);
+
+        redirectAfterSuccess(req, res);
+    }
+
+    @Override
+    protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
+        AuthenticationException failed) throws IOException, ServletException {
+
+        String baseRediredirectUrl = configurationService.getProperty("dspace.ui.url");
+        String redirectUrl = baseRediredirectUrl + "/error?status=401&code=orcid.generic-error";
+        response.sendRedirect(redirectUrl);
+
+    }
+
+    /**
+     * After successful login, redirect to the DSpace URL specified by this Orcid
+     * request (in the "redirectUrl" request parameter). If that 'redirectUrl' is
+     * not valid or trusted for this DSpace site, then return a 400 error.
+     * @param  request
+     * @param  response
+     * @throws IOException
+     */
+    private void redirectAfterSuccess(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        // Get redirect URL from request parameter
+        String redirectUrl = request.getParameter("redirectUrl");
+
+        // If redirectUrl unspecified, default to the configured UI
+        if (StringUtils.isEmpty(redirectUrl)) {
+            redirectUrl = configurationService.getProperty("dspace.ui.url");
+        }
+
+        // Validate that the redirectURL matches either the server or UI hostname. It
+        // *cannot* be an arbitrary URL.
+        String redirectHostName = Utils.getHostName(redirectUrl);
+        String serverHostName = Utils.getHostName(configurationService.getProperty("dspace.server.url"));
+        ArrayList<String> allowedHostNames = new ArrayList<>();
+        allowedHostNames.add(serverHostName);
+        String[] allowedUrls = configurationService.getArrayProperty("rest.cors.allowed-origins");
+        for (String url : allowedUrls) {
+            allowedHostNames.add(Utils.getHostName(url));
+        }
+
+        if (StringUtils.equalsAnyIgnoreCase(redirectHostName, allowedHostNames.toArray(new String[0]))) {
+            log.debug("Orcid redirecting to " + redirectUrl);
+            response.sendRedirect(redirectUrl);
```

Comment:
To silence the false-positive warning from LGTM, append this on the end of this line of code:
`// lgtm [java/unvalidated-url-redirection]`

We had to add this same thing to the OIDC redirect (another false positive): https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/OidcRestController.java#L70